### PR TITLE
#5489 Mock the metrics data for fake connector and topic

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist

- [x] [Review Contribution Guidelines](https://oharastream.github.io/en/docs/master/contributing/)
- [x] Add at least one reviewer (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add an assignee (of course, you are the assignee by default)
- [x] Do not set milestone, projects and labels (PR labels are added automatically by Jenkins)
- [x] Link to an issue
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (If the failed tests are unrelated to this PR, add a comment in the PR thread)

### CODE REVIEW

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 1]
```diff
@@ -139,10 +143,22 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
...
+
   override protected def topicMeters(cluster: ClusterInfo): Map[String, Seq[TopicMeter]] = cluster match {
     case _: BrokerClusterInfo =>
-      // we don't care for the fake mode since both fake mode and embedded mode are run on local jvm
-      Map(CommonUtils.hostname() -> BeanChannel.local().topicMeters().asScala.toSeq)
+      if (clusterCache.containsKey(cluster.key)) {
+        import scala.concurrent.ExecutionContext.Implicits.global
+        val topicKeys =
+          Await.result(dataCollie.values[TopicInfo]().map(value => value.map(info => info.key)), Duration.Inf)
+        topicKeys.map(topicKey => CommonUtils.hostname() -> Seq(fakeTopicMeter(topicKey))).toMap
```
Comment (saivirtue) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:160]: In this scope, we could not find the "topic keys" by clusterInfo. So I fetch the raw data to achieve this instead.
It's a workaround for fake mode of configurator to show the metric data.

Comment (chia7712) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:160]: 這邊要處理兩種情境，一種是fake (no server)，一種是embedded

現在的改法會讓embedded下的“真”metrics都變成fake的

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 2]
```diff
@@ -139,10 +143,22 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
...
+      .build()
+
   override protected def topicMeters(cluster: ClusterInfo): Map[String, Seq[TopicMeter]] = cluster match {
     case _: BrokerClusterInfo =>
-      // we don't care for the fake mode since both fake mode and embedded mode are run on local jvm
-      Map(CommonUtils.hostname() -> BeanChannel.local().topicMeters().asScala.toSeq)
+      if (clusterCache.containsKey(cluster.key)) {
+        import scala.concurrent.ExecutionContext.Implicits.global
+        val topicKeys =
+          Await.result(dataCollie.values[TopicInfo]().map(value => value.map(info => info.key)), Duration.Inf)
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:159]: 另外這邊應該要只為"running"的topics產生metrics

從dataCollie拿來的topics並不一定都是running

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 3]
```diff
@@ -165,6 +181,13 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
       if (clusterCache.containsKey(cluster.key)) {
         Map(CommonUtils.hostname() -> Seq(fakeCounter(cluster.key)))
       } else Map.empty
+    case _ @(_: WorkerClusterInfo) =>
+      if (clusterCache.containsKey(cluster.key)) {
+        import scala.concurrent.ExecutionContext.Implicits.global
+        val connectorKeys =
+          Await.result(dataCollie.values[ConnectorInfo]().map(value => value.map(info => info.key)), Duration.Inf)
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:188]: 這邊也有類似的狀況

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala (hunk 1 above)
Comment (saivirtue) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:160]: Just curious, do we have such a situation that "using embedded metrics with fake configurator" ?

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala (hunk 2 above)
Comment (saivirtue) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:159]: Good point.

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala (hunk 1 above)
Comment (chia7712) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:160]: The metrics test had verified the number of topic. It does not care for the accurate number so this PR should work. However, this PR still broke something that the number of metrics from embedded mode is not accurate.

**Comment by saivirtue** [comment]
retry collieIT

**Comment by saivirtue** [comment]
retry connectorIT

**Comment by saivirtue** [comment]
retry otherIT

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 4]
```diff
@@ -139,10 +143,32 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
...
+      .topicName(key.toPlain)
+      .catalog(TopicMeter.Catalog.MessagesInPerSec)
+      .rateUnit(TimeUnit.SECONDS)
+      .build()
+
   override protected def topicMeters(cluster: ClusterInfo): Map[String, Seq[TopicMeter]] = cluster match {
     case _: BrokerClusterInfo =>
-      // we don't care for the fake mode since both fake mode and embedded mode are run on local jvm
-      Map(CommonUtils.hostname() -> BeanChannel.local().topicMeters().asScala.toSeq)
+      if (!BeanChannel.local().topicMeters().isEmpty) {
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:156]: 這個判斷式可否獨立成一個method，例如：
```
def isEmbedded: Boolean = BeanChannel.local().topicMeters().isEmpty
```

然後```FakeBrokerCollie``` and ```FakeWorkerCollie```應該重寫該方法改成用有沒有```bkConnectionProps```/```wkConnectionProps```來判斷

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 5]
```diff
@@ -165,6 +191,25 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
       if (clusterCache.containsKey(cluster.key)) {
         Map(CommonUtils.hostname() -> Seq(fakeCounter(cluster.key)))
       } else Map.empty
+    case _ @(_: WorkerClusterInfo) =>
+      if (!BeanChannel.local().counterMBeans().isEmpty) {
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:195]: 同上

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala (hunk 4 above)
Comment (saivirtue) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:156]: Sure

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala [hunk 6]
```diff
@@ -81,6 +85,30 @@ private[configurator] abstract class FakeCollie(val dataCollie: DataCollie) exte
       )
     )
 
+  /**
+    * Test this collie is running on embedded mode or not by checking local JVM metrics.
+    * @return true if run on embedded mode
+    */
+  private[configurator] def isEmbedded: Boolean = BeanChannel.local().nonEmpty()
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:92]: 考量到之後Configurator可能會也有metrics，是否額外判斷是否來自topic or connector? 就用你上隻commit的方法

Comment (saivirtue) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/fake/FakeCollie.scala:92]: 會改成判斷全部主要是因為考慮到未來有新的metric type也會需要在這裡加入判斷，覺得有點不方便。
但還是先以目前有的來考慮好了

**Comment by saivirtue** [comment]
retry clientIT

**Comment by saivirtue** [comment]
retry connectorIT

**Comment by saivirtue** [comment]
retry otherIT

**Comment by saivirtue** [comment]
retry streamIT

**Comment by saivirtue** [comment]
retry connector

**Pull request merged**